lib/IDE/CodeCompletion.cpp

```diff
@@ -4124,6 +4126,79 @@ class CompletionLookup final : public swift::VisibleDeclConsumer {
     }
   }
 
+  void collectPrecedenceGroups() {
+    assert(CurrDeclContext);
+
+    auto addPrecedenceGroup = [&](PrecedenceGroupDecl *PGD) {
+      auto semanticContext =
+        getSemanticContext(PGD, DeclVisibilityKind::VisibleAtTopLevel);
+      CodeCompletionResultBuilder builder(
+        Sink, CodeCompletionResult::ResultKind::Declaration,
+        semanticContext, {});
+      builder.addTextChunk(PGD->getName().str());
+      builder.setAssociatedDecl(PGD);
+    };
+
+    auto collectPrecedenceGroups = [&](FileUnit *FU) {
+      switch (FU->getKind()) {
+      case FileUnitKind::SerializedAST: {
+        SmallVector<Decl *, 16> topLevelDecls;
+        cast<LoadedFile>(FU)->getTopLevelDecls(topLevelDecls);
+        for (auto D: topLevelDecls)
+          if (auto PGD = dyn_cast<PrecedenceGroupDecl>(D))
+            addPrecedenceGroup(PGD);
+        break;
+      }
+      case FileUnitKind::Source: {
+        auto file = cast<SourceFile>(FU);
+        bool includePrivate = CurrDeclContext->getParentSourceFile() == file;
+        for (auto &pair : file->PrecedenceGroups) {
+          PrecedenceGroupDecl *group;
+          if ((group = pair.second.getPointer()) &&
+              (pair.second.getInt() || includePrivate))
+            addPrecedenceGroup(group);
+        }
+        break;
+      }
+      default: return;
+      }
+    };
+
+    CurrDeclContext->getParentSourceFile()->
+      forAllVisibleModules([&](ModuleDecl::ImportedModule import) {
+       for (auto fileUnit: import.second->getFiles())
+          collectPrecedenceGroups(fileUnit);
+    });
+  }
+
+  void getPrecedenceGroupCompletions(SyntaxKind SK) {
+#define AddKeyword(kw) addKeyword(getTokenText(tok::kw), None,    \
```

Comment:
Could you make it a closure or function call instead of a macro for debuggability?

Reply:
If twice as much space wasted is not a problem – of course.

Reply:
Yeah, space should be fine. Readability and debuggability are more important.